File: backend/src/main/java/com/bablsoft/accessflow/core/api/SystemRolePermissions.java
package com.bablsoft.accessflow.core.api;

import java.util.Collections;
import java.util.EnumMap;
import java.util.EnumSet;
import java.util.Map;
import java.util.Set;

/**
 * The authoritative permission sets of the 5 immutable system roles (AF-522), reproducing the
 * pre-custom-roles authorization matrix exactly. Runtime resolution answers system roles from this
 * map; the {@code V114} seed rows mirror it for catalog/UI purposes and a parity test keeps the
 * two in sync.
 */
public final class SystemRolePermissions {

    private static final Map<UserRoleType, Set<Permission>> SETS = buildSets();

    private SystemRolePermissions() {
    }

    /** The permission set of the given system role; never null. */
    public static Set<Permission> of(UserRoleType role) {
        return SETS.get(role);
    }

    private static Map<UserRoleType, Set<Permission>> buildSets() {
        var readonly = EnumSet.of(Permission.QUERY_SUBMIT_SELECT);

        var analyst = EnumSet.copyOf(readonly);
        analyst.add(Permission.QUERY_SUBMIT_DML);

        var reviewer = EnumSet.copyOf(analyst);
        reviewer.add(Permission.QUERY_VIEW_ALL);
        reviewer.add(Permission.QUERY_REVIEW);
        reviewer.add(Permission.ACCESS_REQUEST_REVIEW);
        reviewer.add(Permission.API_REQUEST_REVIEW);
        reviewer.add(Permission.ERASURE_REVIEW);
        reviewer.add(Permission.ATTESTATION_REVIEW);

        var auditor = EnumSet.of(
                Permission.COMPLIANCE_REPORT_VIEW,
                Permission.ATTESTATION_EVIDENCE_EXPORT,
                Permission.BREAK_GLASS_VIEW,
                Permission.ANOMALY_VIEW);

        var admin = EnumSet.allOf(Permission.class);

        Map<UserRoleType, Set<Permission>> sets = new EnumMap<>(UserRoleType.class);
        sets.put(UserRoleType.READONLY, Collections.unmodifiableSet(readonly));
        sets.put(UserRoleType.ANALYST, Collections.unmodifiableSet(analyst));
        sets.put(UserRoleType.REVIEWER, Collections.unmodifiableSet(reviewer));
        sets.put(UserRoleType.AUDITOR, Collections.unmodifiableSet(auditor));
        sets.put(UserRoleType.ADMIN, Collections.unmodifiableSet(admin));
        return Collections.unmodifiableMap(sets);
    }
}

File: frontend/src/mocks/systemRolePermissions.ts
import { PERMISSIONS, type Permission } from '@/utils/permissions';

/**
 * Test fixture (AF-522): representative permission sets for the built-in system roles,
 * mirroring the backend's seeded role → permission mapping closely enough for UI gating.
 * Production code never uses this — gating always reads the user's `permissions` array.
 */
export const SYSTEM_ROLE_PERMISSIONS: Record<
  'ADMIN' | 'REVIEWER' | 'ANALYST' | 'READONLY' | 'AUDITOR',
  Permission[]
> = {
  ADMIN: [...PERMISSIONS],
  REVIEWER: [
    'QUERY_SUBMIT_SELECT',
    'QUERY_SUBMIT_DML',
    'QUERY_VIEW_ALL',
    'QUERY_REVIEW',
    'API_REQUEST_REVIEW',
    'ACCESS_REQUEST_REVIEW',
    'ATTESTATION_REVIEW',
    'ERASURE_REVIEW',
  ],
  ANALYST: ['QUERY_SUBMIT_SELECT', 'QUERY_SUBMIT_DML'],
  READONLY: ['QUERY_SUBMIT_SELECT'],
  AUDITOR: [
    'COMPLIANCE_REPORT_VIEW',
    'AUDIT_LOG_VIEW',
    'BREAK_GLASS_VIEW',
    'ANOMALY_VIEW',
    'ATTESTATION_EVIDENCE_EXPORT',
  ],
};
